// koa-ts-controllers/dist/tests/bootstrap.flow.inheritance.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const supertest_1 = __importDefault(require("supertest"));
const koa_1 = __importDefault(require("koa"));
const index_1 = require("../index");
const flow_1 = require("./util/flow/flow");
let app;
let nativeServer;
let testServer;
beforeAll(async () => {
    app = new koa_1.default();
    await index_1.bootstrapControllers(app, {
        basePath: '/api',
        controllers: [__dirname + '/util/controllers/**/*.ts'],
        boomifyErrors: true,
        initBodyParser: true,
        versions: ['1', '2'],
        flow: [flow_1.setSomethingStateFlow]
    });
    nativeServer = app.listen();
    testServer = supertest_1.default(nativeServer);
});
afterAll((done) => {
    if (nativeServer.listening) {
        nativeServer.close(done);
    }
    else {
        done();
    }
});
describe('Bootstrap function', () => {
    describe('Flow', () => {
        it('works', async () => {
            const response = await testServer
                .get('/api/v2/action/staten')
                .expect(200);
            expect(response.body.something).toEqual('hahaha');
        });
    });
});
